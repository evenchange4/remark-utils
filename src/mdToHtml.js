// @flow
import remark from 'remark';
import remarkGridTables from 'remark-grid-tables';
import remarkHtml from 'remark-html';
import remarkPrismjs from 'gatsby-remark-prismjs';
import remarkAutolink from 'gatsby-remark-autolink-headers';
import defaultOptions from './utils/defaultOptions';
import remarkLazysizes from './utils/remark-lazysizes';
import { type Option } from './utils/type.flow';

const mdToHtml = (markdownContent: string, options?: Option): string => {
  const { autolinkHeaders, prismjs, lazysizes } = {
    ...defaultOptions,
    ...options,
  };

  return remark()
    .use(remarkGridTables)
    .use(() => markdownAST => remarkAutolink({ markdownAST }, autolinkHeaders))
    .use(() => markdownAST => remarkPrismjs({ markdownAST }, prismjs))
    .use(remarkLazysizes, lazysizes)
    .use(remarkHtml)
    .processSync(markdownContent).contents;
};

export default mdToHtml;
