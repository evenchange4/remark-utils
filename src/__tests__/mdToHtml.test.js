// @flow
import mdToHtml from '../mdToHtml';

it('Heading with anchor', () => {
  expect(
    mdToHtml(`
# header 1 標頭

## header 2 標頭

### header 3 標頭

#### header 4 標頭

##### header 5 標頭

###### header 6 標頭
`),
  ).toMatchSnapshot();
});

it('Link', () => {
  expect(
    mdToHtml(`
[react-component](./react-component)

[連結文字](http://dev.nodeca.com)

[加上標題的連結文字](http://nodeca.github.io/pica/demo/ '標題文字！')

自動轉換連結 https://github.com/nodeca/pica
`),
  ).toMatchSnapshot();
});

it('List', () => {
  expect(
    mdToHtml(`
- 在行開頭使用 \`+\` \`-\` 或是 \`*\` 來建立清單
- 空兩個空白就可以產生子清單
  - 當清單標記使用的字元不同，會強制建立新的清單
    - Ac tristique libero volutpat at
    * Facilisis in pretium nisl aliquet
    - Nulla volutpat aliquam velit
- 非常簡單！

1. Lorem ipsum dolor sit amet
2. Consectetur adipiscing elit
3. Integer molestie lorem at massa

57. foo
58. bar
`),
  ).toMatchSnapshot();
});

it('Image', () => {
  expect(
    mdToHtml(`
![AWS_Icons-300x200.png](../images/AWS_Icons-300x200.png 'aws')

![Alt text][id]
使用參考，可以在稍後的文件中再定義圖片網址

[id]: https://octodex.github.com/images/dojocat.jpg 'The Dojocat'
`),
  ).toMatchSnapshot();
});

it('code', () => {
  expect(
    mdToHtml(`
\`\`\`python
# This program adds two numbers

num1 = 1.5
num2 = 6.3

# Add two numbers
sum = float(num1) + float(num2)

# Display the sum
print('The sum of {0} and {1} is {2}'.format(num1, num2, sum))
\`\`\`
`),
  ).toMatchSnapshot();
});

it('quote', () => {
  expect(
    mdToHtml(`
> Tip
>
> Whenever you get confused by something in JavaScript, [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript) and [javascript.info](http://javascript.info/) are great websites to check. There are also [community support forums](/community/support.html) where you can ask for help.

> 引用區塊也可以是巢狀的喔...
>
> > ...可以多層次的使用...
> >
> > > ...或是用空白隔開
`),
  ).toMatchSnapshot();
});

it('table native', () => {
  expect(
    mdToHtml(`
| 選項   | 描述                                                                      |
| ------ | ------------------------------------------------------------------------- |
| data   | path to data files to supply the data that will be passed into templates. |
| engine | engine to be used for processing templates. Handlebars is the default.    |
| ext    | extension to be used for dest files.                                      |
`),
  ).toMatchSnapshot();
});

it('table extend', () => {
  expect(
    mdToHtml(`
+-------+----------+------+
| Table Headings   | Here |
+-------+----------+------+
| Sub   | Headings | Too  |
+=======+==========+======+
| cell  | column spanning |
+ spans +----------+------+
| rows  | normal   | cell |
+-------+----------+------+
| multi | cells can be    |
| line  | *formatted*     |
|       | **paragraphs**  |
| cells |                 |
| too   |                 |
+-------+-----------------+
`),
  ).toMatchSnapshot();
});

it('font style', () => {
  expect(
    mdToHtml(`
**這是粗體文字**

_這是斜體文字_

~~這是刪除文字~~
`),
  ).toMatchSnapshot();
});
