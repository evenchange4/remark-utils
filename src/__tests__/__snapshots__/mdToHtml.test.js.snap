// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading with anchor 1`] = `
"<h1 id=\\"header-1-標頭\\"><a href=\\"#header-1-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 1 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 1 標頭</h1>
<h2 id=\\"header-2-標頭\\"><a href=\\"#header-2-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 2 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 2 標頭</h2>
<h3 id=\\"header-3-標頭\\"><a href=\\"#header-3-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 3 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 3 標頭</h3>
<h4 id=\\"header-4-標頭\\"><a href=\\"#header-4-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 4 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 4 標頭</h4>
<h5 id=\\"header-5-標頭\\"><a href=\\"#header-5-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 5 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 5 標頭</h5>
<h6 id=\\"header-6-標頭\\"><a href=\\"#header-6-%E6%A8%99%E9%A0%AD\\" aria-label=\\"header 6 標頭 permalink\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" height=\\"24\\" version=\\"1.1\\" viewBox=\\"0 0 24 24\\" width=\\"24\\"><path fill-rule=\\"evenodd\\" d=\\"M3.9,12c0-1.7,1.4-3.1,3.1-3.1h4V7H7c-2.8,0-5,2.2-5,5s2.2,5,5,5h4v-1.9H7C5.3,15.1,3.9,13.7,3.9,12z M8,13h8v-2H8V13z M17,7h-4v1.9h4c1.7,0,3.1,1.4,3.1,3.1s-1.4,3.1-3.1,3.1h-4V17h4c2.8,0,5-2.2,5-5S19.8,7,17,7z\\"></path></svg></a>header 6 標頭</h6>
"
`;

exports[`Image 1`] = `
"<p><img src=\\"../images/AWS_Icons-300x200.png\\" alt=\\"AWS_Icons-300x200.png\\" title=\\"aws\\"></p>
<p><img src=\\"https://octodex.github.com/images/dojocat.jpg\\" alt=\\"Alt text\\" title=\\"The Dojocat\\">
使用參考，可以在稍後的文件中再定義圖片網址</p>
"
`;

exports[`Image with lazysizes 1`] = `
"<p><div style=\\"max-width: 300px;\\"><div style=\\"width: 100%; padding-bottom: 66.66666666666666%; height: 0; position: relative;\\"><img src=\\"data:image/png;base64,mock\\" alt=\\"AWS_Icons-300x200.png\\" title=\\"aws\\" data-src=\\"../images/AWS_Icons-300x200.png\\" class=\\"lazyload\\" style=\\"position: absolute; width: 100%;\\"></div></div>
<img src=\\"../images/react.png\\" alt=\\"react.png\\" title=\\"react\\"></p>
<p><img src=\\"https://octodex.github.com/images/dojocat.jpg\\" alt=\\"Alt text\\" title=\\"The Dojocat\\">
使用參考，可以在稍後的文件中再定義圖片網址</p>
"
`;

exports[`Link 1`] = `
"<p><a href=\\"./react-component\\">react-component</a></p>
<p><a href=\\"http://dev.nodeca.com\\">連結文字</a></p>
<p><a href=\\"http://nodeca.github.io/pica/demo/\\" title=\\"標題文字！\\">加上標題的連結文字</a></p>
<p>自動轉換連結 <a href=\\"https://github.com/nodeca/pica\\">https://github.com/nodeca/pica</a></p>
"
`;

exports[`List 1`] = `
"<ul>
<li>在行開頭使用 <code class=\\"language-text\\">+</code> <code class=\\"language-text\\">-</code> 或是 <code class=\\"language-text\\">*</code> 來建立清單</li>
<li>空兩個空白就可以產生子清單<ul>
<li>當清單標記使用的字元不同，會強制建立新的清單<ul>
<li>Ac tristique libero volutpat at</li>
</ul><ul>
<li>Facilisis in pretium nisl aliquet</li>
</ul><ul>
<li>Nulla volutpat aliquam velit</li>
</ul></li>
</ul></li>
<li>非常簡單！</li>
</ul>
<ol>
<li>
<p>Lorem ipsum dolor sit amet</p>
</li>
<li>
<p>Consectetur adipiscing elit</p>
</li>
<li>
<p>Integer molestie lorem at massa</p>
</li>
<li>
<p>foo</p>
</li>
<li>
<p>bar</p>
</li>
</ol>
"
`;

exports[`code 1`] = `
"<div class=\\"gatsby-highlight\\" data-language=\\"python\\"><pre class=\\"language-python\\"><code class=\\"language-python\\"><span class=\\"token comment\\"># This program adds two numbers</span>

num1 <span class=\\"token operator\\">=</span> <span class=\\"token number\\">1.5</span>
num2 <span class=\\"token operator\\">=</span> <span class=\\"token number\\">6.3</span>

<span class=\\"token comment\\"># Add two numbers</span>
<span class=\\"token builtin\\">sum</span> <span class=\\"token operator\\">=</span> <span class=\\"token builtin\\">float</span><span class=\\"token punctuation\\">(</span>num1<span class=\\"token punctuation\\">)</span> <span class=\\"token operator\\">+</span> <span class=\\"token builtin\\">float</span><span class=\\"token punctuation\\">(</span>num2<span class=\\"token punctuation\\">)</span>

<span class=\\"token comment\\"># Display the sum</span>
<span class=\\"token keyword\\">print</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">'The sum of {0} and {1} is {2}'</span><span class=\\"token punctuation\\">.</span><span class=\\"token builtin\\">format</span><span class=\\"token punctuation\\">(</span>num1<span class=\\"token punctuation\\">,</span> num2<span class=\\"token punctuation\\">,</span> <span class=\\"token builtin\\">sum</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">)</span></code></pre></div>
"
`;

exports[`font style 1`] = `
"<p><strong>這是粗體文字</strong></p>
<p><em>這是斜體文字</em></p>
<p><del>這是刪除文字</del></p>
"
`;

exports[`quote 1`] = `
"<blockquote>
<p>Tip</p>
<p>Whenever you get confused by something in JavaScript, <a href=\\"https://developer.mozilla.org/en-US/docs/Web/JavaScript\\">MDN</a> and <a href=\\"http://javascript.info/\\">javascript.info</a> are great websites to check. There are also <a href=\\"/community/support.html\\">community support forums</a> where you can ask for help.</p>
<p>引用區塊也可以是巢狀的喔...</p>
<blockquote>
<p>...可以多層次的使用...</p>
<blockquote>
<p>...或是用空白隔開</p>
</blockquote>
</blockquote>
</blockquote>
"
`;

exports[`table extend 1`] = `
"<table><thead><tr><th colspan=\\"2\\" rowspan=\\"1\\"><p>Table Headings</p></th><th colspan=\\"1\\" rowspan=\\"1\\"><p>Here</p></th></tr><tr><th colspan=\\"1\\" rowspan=\\"1\\"><p>Sub</p></th><th colspan=\\"1\\" rowspan=\\"1\\"><p>Headings</p></th><th colspan=\\"1\\" rowspan=\\"1\\"><p>Too</p></th></tr></thead><tbody><tr><td colspan=\\"1\\" rowspan=\\"2\\"><p>cell
spans
rows</p></td><td colspan=\\"2\\" rowspan=\\"1\\"><p>column spanning</p></td></tr><tr><td colspan=\\"1\\" rowspan=\\"1\\"><p>normal</p></td><td colspan=\\"1\\" rowspan=\\"1\\"><p>cell</p></td></tr><tr><td colspan=\\"1\\" rowspan=\\"1\\"><p>multi
line</p><p>cells
too</p></td><td colspan=\\"2\\" rowspan=\\"1\\"><p>cells can be
<em>formatted</em>
<strong>paragraphs</strong></p></td></tr></tbody></table>
"
`;

exports[`table native 1`] = `
"<table>
<thead>
<tr>
<th>選項</th>
<th>描述</th>
</tr>
</thead>
<tbody>
<tr>
<td>data</td>
<td>path to data files to supply the data that will be passed into templates.</td>
</tr>
<tr>
<td>engine</td>
<td>engine to be used for processing templates. Handlebars is the default.</td>
</tr>
<tr>
<td>ext</td>
<td>extension to be used for dest files.</td>
</tr>
</tbody>
</table>
"
`;
